### PR TITLE
section ADDR_PERSISTENT defined in FLASH

### LINKED ISSUE #34: investigate linker script with reserved flash space
find way to reserve block(s) of flash space, so that it can't be overwritten by new, larger, firmware

- [x] - look for existing projects with custom linker / loader script
- [x] - learn
- [x] - make proof of concept
- [x] - adapt to code base

[sdk doc with linker info](https://datasheets.raspberrypi.com/pico/raspberry-pi-pico-c-sdk.pdf).

```
pico_standard_link encapsulates the standard linker setup needed to configure the type of application binary layout in
memory, and link to any additional C and/or C++ runtime libraries. It also includes the default crt0, which provides the
initial entry point from the flash second stage bootloader, contains the initial vector table (later relocated to RAM), and
initialises static data and RAM-resident code if the application is running from flash.
```

[custom linker file](https://github.com/rhulme/pico-flashloader/blob/master/CMakeLists.txt):
`pico_set_linker_script(my_target ${CMAKE_CURRENT_LIST_DIR}/custom.ld)`

[default loader source](https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/pico_standard_link/memmap_default.ld).

[sdk linker settings](https://github.com/search?q=repo%3Araspberrypi%2Fpico-sdk%20PICO_COPY_TO_RAM&type=code)  

[Proof of concept](https://community.element14.com/products/raspberry-pi/b/blog/posts/raspberry-pico-c-sdk-reserve-a-flash-memory-block-for-persistent-storage)

@shabaz123 , I am splitting this part out and assigning to self. 
I think we can build the flash stgorage part and the memory protection part separately,
integrate the pieces that work, when they work.

### PR DESCRIPTION
**Pull request by jancumps**
created dedicated flash area ADDR_PERSISTENT
4K
created a definition, so that changing the size requires no calculations

closes #34 

to use the address of this area in C (in flash API):

```
       extern uint32_t ADDR_PERSISTENT[];
       #define ADDR_PERSISTENT_BASE_ADDR   (ADDR_PERSISTENT)    
```

### CODE REVIEW

File: memmap_custom.ld [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    __bss_end__
+    __end__
+    end
+    __HeapLimit
+    __StackLimit
+    __StackTop
+    __stack (== StackTop)
+*/
+
+/* PST custom: claim a persistent storage area from FLASH */
```
Comment (jancumps) [inline memmap_custom.ld:24]: this is the first change to the default script: make flash smaller, add persistent area

File: memmap_custom.ld [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+        KEEP(*(.binary_info.keep.*))
+        *(.binary_info.*)
+    } > FLASH
+    __binary_info_end = .;
+    . = ALIGN(4);
+
+    /* End of .text-like segments */
+    __etext = .;
+    
+    /* PST custom: define symbol for persistent area that
```
Comment (jancumps) [inline memmap_custom.ld:127]: 2nd change to the default script. Define the symbol that can be used in c code

File: CMakeLists.txt [hunk 3]
```diff
@@ -54,6 +54,9 @@ target_include_directories(pico_scpi_usbtmc_labtool PRIVATE
 )
 
 target_link_libraries(pico_scpi_usbtmc_labtool PUBLIC pico_stdlib tinyusb_device tinyusb_board hardware_gpio hardware_adc hardware_pwm hardware_i2c pico_unique_id)
+
+pico_set_linker_script(pico_scpi_usbtmc_labtool ${CMAKE_SOURCE_DIR}/memmap_custom.ld)
```
Comment (jancumps) [inline CMakeLists.txt:58]: use our custom script instead of default one fro c sdk

**Review by shabaz123 — APPROVED** [review]
Looks good!

**Pull request merged**